syntax = "proto3"; // we use proto version 3
package consensus_interface;   // similar to namespace, avoid conflicts

/** Paxos consensus **/
service ConsensusService { 
  rpc propose(Request) returns (Response) {};
  rpc accept(Request) returns (Response) {};
  rpc success(Request) returns (Response) {};
  rpc ping(Empty) returns (Empty) {};
  rpc get_leader(Empty) returns (GetLeaderResponse) {};
  rpc elect_leader(ElectLeaderRequest) returns (Empty) {};
}

message Empty {

}


enum Operation {
  GET = 0;
  SET = 1;
  DELETE = 2;
  SET_LEADER = 3;
  NOT_SET = 4;
};

message Request {
    string key = 1;
    string value = 2;
    int32 pserver_id = 3;
    int32 aserver_id = 4;
    int32 round = 5;
    Operation op = 6;
}

enum Status_Types {
   OK = 0;
   FAILED = 1;
 }

message Response {
    string value = 1;
    int32 pserver_id = 2;
    int32 aserver_id = 3;
    int32 round = 4;
    Operation op = 5;
    Status_Types status = 6;
}

message GetLeaderResponse{
  string leader = 1;
}

message ElectLeaderRequest {
	string key = 1;
	string value = 2;
}